_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+  the `plan` output. This `plan` is stored in plan file format so that when upon merge, the exact plan is applied using
+  `terraform apply`. Note that Atlantis is optimized for this single workflow. That is, it cannot be made to implement
+  other workflows, such as building images or running infrastructure tests using terratest. Being a self-hosted
+  platform, you can keep your credentials internal to your data center and lock down server level access. However,
+  Atlantis requires a public facing endpoint to be available so that it can listen on webhooks from the major VCS
+  platforms (GitHub/GitLab/Bitbucket/Azure DevOps). This means that you are relying on the security model of the
+  Atlantis platform to ensure that it doesn't accidentally run infrastructure code from unwanted sources with the
+  credentials of the platform. To prevent unwanted deployments, you will want to ensure that you are following
+  https://www.runatlantis.io/docs/security.html#mitigations[all the security best practices] of the platform.
+
+https://www.hashicorp.com/products/terraform/[Terraform Enterprise and Terraform Cloud] (self-hosted and Saas, specialized)::
+  Terraform Cloud (Saas) and Terraform Enterprise (self-hosted) are platforms provided by HashiCorp (the creator of
+  Terraform) that are optimized for Terraform workflows. Both services provide a web UI for remotely running `terraform
+  plan` and `terraform apply`, and can be integrated with VCS platforms to implement CI/CD workflows that remotely run
+  `terraform` in reaction to git. Additionally, Terraform Enterprise in particular provides the ability to manage
+  deployment policies and permissions as code. Note that Terraform Cloud and Enterprise are both optimized for running
+  `terraform`. That is, they do not support the use of external binaries such as `terragrunt` or `docker`, even when
+  called from within `terraform` using a `local-exec` provisioner or external data source. In terms of the threat model,
+  Terraform Cloud and Enterprise both implement respectable security practices, including end to end encryption backed
+  by HashiCorp Vault, and you have the option to run in your own data center with Terraform Enterprise to lock it down.
+  Note that Terraform Enterprise relies on API based integrations with the VCS platforms instead of webhooks, so it is
+  not necessary to have a publicly facing service when running in self-hosted mode.
+
+
+[[production_grade_design]]
+== Production-grade design
+
+With all the core concepts out of the way, let's now discuss how to configure a production-grade CI/CD workflow for
+infrastructure code, using a platform that looks something like this:
+
+.Architecture of platform for running Terraform/Terragrunt CI/CD workflows.
+image::/assets/img/guides/infrastructure-cicd-pipeline/tftg-pipeline-architecture.png[]
+
+.Sequence diagram of running Terraform/Terragrunt CI/CD workflows.
```

Comment:
Might want to save this sequence diagram until later. Otherwise, two large, complicated images in a row can be a lot of info to take in.

Reply:
Agreed. The architecture diagram followed immediately by the sequence diagram is intense.

Reply:
Moved 6a706196717b2ebced13ed710010ed0974ed9353